### ISSUE TITLE
Sparse Inputs

### ISSUE THREAD

**Issue by adodge** [hyperopt/hyperopt-sklearn#105]
Hi!  I'm having a problem with using sparse inputs.  Here's an example which reproduces the issue:

```
#!/usr/bin/env python

from hpsklearn import HyperoptEstimator, sgd
import scipy.sparse as ss
import numpy as np

# Generate some random sparse data
nrows,ncols,nnz = 100,50,10
ntrue = nrows // 2
D,C,R = [],[],[]
for r in range(nrows):
    feats = np.random.choice(range(ncols), size=nnz, replace=False)
    D.extend([1]*nnz)
    C.extend(feats)
    R.extend([r]*nnz)
X = ss.csr_matrix( (D,(R,C)), shape=(nrows, ncols))
y = np.zeros( nrows )
y[:ntrue] = 1


# Try to fit an SGD model
cls = HyperoptEstimator(
    classifier=sgd('sgd', loss='log'),
    preprocessing=[],
)
cls.fit(X,y)
```

```
Traceback (most recent call last):
  File "problem.py", line 26, in <module>
    cls.fit(X,y)
  File "/home/adodge/dev/oss/hyperopt/hyperopt-sklearn/hpsklearn/estimator.py", line 747, in fit
    fit_iter.send(increment)
  File "/home/adodge/dev/oss/hyperopt/hyperopt-sklearn/hpsklearn/estimator.py", line 657, in fit_iter
    return_argmin=False, # -- in case no success so far
  File "/home/adodge/dev/oss/hyperopt/env/lib/python3.6/site-packages/hyperopt/fmin.py", line 367, in fmin
    return_argmin=return_argmin,
  File "/home/adodge/dev/oss/hyperopt/env/lib/python3.6/site-packages/hyperopt/base.py", line 635, in fmin
    return_argmin=return_argmin)
  File "/home/adodge/dev/oss/hyperopt/env/lib/python3.6/site-packages/hyperopt/fmin.py", line 385, in fmin
    rval.exhaust()
  File "/home/adodge/dev/oss/hyperopt/env/lib/python3.6/site-packages/hyperopt/fmin.py", line 244, in exhaust
    self.run(self.max_evals - n_done, block_until_done=self.asynchronous)
  File "/home/adodge/dev/oss/hyperopt/env/lib/python3.6/site-packages/hyperopt/fmin.py", line 218, in run
    self.serial_evaluate()
  File "/home/adodge/dev/oss/hyperopt/env/lib/python3.6/site-packages/hyperopt/fmin.py", line 137, in serial_evaluate
    result = self.domain.evaluate(spec, ctrl)
  File "/home/adodge/dev/oss/hyperopt/env/lib/python3.6/site-packages/hyperopt/base.py", line 840, in evaluate
    rval = self.fn(pyll_rval)
  File "/home/adodge/dev/oss/hyperopt/hyperopt-sklearn/hpsklearn/estimator.py", line 620, in fn_with_timeout
    raise fn_rval[1]
ValueError: zero-dimensional arrays cannot be concatenated
```

By inserting some traceroute calls inside the estimator code, I see the problem is on line 122:

```
122         XEXfit = np.concatenate(XEXfit_list, axis=1)
```

There are several np.concatenate calls here, any of which will fail if given a sparse matrix.  I can fix this with the following patch:

```
--- a/hpsklearn/estimator.py
+++ b/hpsklearn/estimator.py
@@ -114,13 +114,22 @@ def transform_combine_XEX(Xfit, info, en_pps=[], Xval=None,
         else:
             transformed_XEX_list.append(dfit)

+    def safe_concatenate(XS):
+        if not any(scipy.sparse.issparse(x) for x in XS):
+            return np.concatenate(XS, axis=1)
+
+        XS = [ x if scipy.sparse.issparse(x) else scipy.sparse.csr_matrix(x)
+               for x in XS ]
+
+        return scipy.sparse.hstack(XS)
+
     if Xval is None:
-        XEXfit = np.concatenate(transformed_XEX_list, axis=1)
+        XEXfit = safe_concatenate(transformed_XEX_list)
         return XEXfit
     else:
         XEXfit_list, XEXval_list = zip(*transformed_XEX_list)
-        XEXfit = np.concatenate(XEXfit_list, axis=1)
-        XEXval = np.concatenate(XEXval_list, axis=1)
+        XEXfit = safe_concatenate(XEXfit_list)
+        XEXval = safe_concatenate(XEXval_list)
         return (XEXfit, XEXval)
```

However!  I can't shake the feeling that the problem here is me, since it's so straightforward to cause this error, maybe I'm set up wrong or I have the wrong expectation for how to call the API.  What do you think?

* python 3.6.3
* numpy 1.15.1
* scipy 1.1.0

Thank you!

**Comment by bjkomer** [comment]
Thanks for the fix and the test in #107! A lot has changed in the codebase since sparse inputs were originally supported, having more tests should hopefully prevent these kinds of errors in the future.

**Comment by stevacca** [comment]
I fixed the _concatenate problem_ but I'm facing a new error: 

Traceback (most recent call last):
  File "C:/Users/steva/PycharmProjects/CreateBOW/prova2.py", line 26, in <module>
    estim.fit(xTrain, yTrain) # error pops up here
  File "C:\Users\steva\PycharmProjects\CreateBOW\venv\lib\site-packages\hpsklearn\estimator.py", line 753, in fit
    fit_iter.send(increment)
  File "C:\Users\steva\PycharmProjects\CreateBOW\venv\lib\site-packages\hpsklearn\estimator.py", line 664, in fit_iter
    return_argmin=False, # -- in case no success so far
  File "C:\Users\steva\PycharmProjects\CreateBOW\venv\lib\site-packages\hyperopt\fmin.py", line 388, in fmin
    show_progressbar=show_progressbar,
  File "C:\Users\steva\PycharmProjects\CreateBOW\venv\lib\site-packages\hyperopt\base.py", line 639, in fmin
    show_progressbar=show_progressbar)
  File "C:\Users\steva\PycharmProjects\CreateBOW\venv\lib\site-packages\hyperopt\fmin.py", line 407, in fmin
    rval.exhaust()
  File "C:\Users\steva\PycharmProjects\CreateBOW\venv\lib\site-packages\hyperopt\fmin.py", line 262, in exhaust
    self.run(self.max_evals - n_done, block_until_done=self.asynchronous)
  File "C:\Users\steva\PycharmProjects\CreateBOW\venv\lib\site-packages\hyperopt\fmin.py", line 227, in run
    self.serial_evaluate()
  File "C:\Users\steva\PycharmProjects\CreateBOW\venv\lib\site-packages\hyperopt\fmin.py", line 141, in serial_evaluate
    result = self.domain.evaluate(spec, ctrl)
  File "C:\Users\steva\PycharmProjects\CreateBOW\venv\lib\site-packages\hyperopt\base.py", line 844, in evaluate
    rval = self.fn(pyll_rval)
  File "C:\Users\steva\PycharmProjects\CreateBOW\venv\lib\site-packages\hpsklearn\estimator.py", line 627, in fn_with_timeout
    raise fn_rval[1]
AttributeError: 'tuple' object has no attribute 'fit'
